Port of multi_margin_loss from TH to ATen (CPU)

andreaskoepf opened the pull request:
This is a port of the existing TH CPU C MultiMarginCriterion to function multi_margin_loss for ATen. ~~The ATen/C++ version is unfortunately significantly slower than the original. It is currently unclear to me what causes the performance degradation since the Tensor access is raw-pointer based similar to the original C implementation. (A first implementation I had created using TensorAccessor was even about 2x slower than the one in this PR).~~

andreaskoepf committed:
Port of multi_margin_loss from TH to ATen (CPU)

First naive try to port the CPU implementation of multi_margin_loss to ATen.
Performace degraded significantly compared to the TH implementation.

andreaskoepf commented:
Note that the benchmark only executes the forward code without reduction, e.g. mainly the following loops: https://github.com/pytorch/pytorch/blob/0249a939afd7bf82ff391fa1e6497a8c56972e88/aten/src/ATen/native/LossMultiMargin.cpp#L10-L56

The corresponding lines in the original TH implementation are here:
https://github.com/pytorch/pytorch/blob/705958be5bda297ed81b6cc29f0f5ef4f3a683ee/aten/src/THNN/generic/MultiMarginCriterion.c#L53-L79

BTW I already tried to manually inline the `multi_margin_inner_sum_cpu()` template function as it is done in the TH-part .. but that had no impact on the performance.

ezyang commented on aten/src/ATen/native/LossMultiMargin.cpp:
```diff
@@ -0,0 +1,328 @@
+  if (reduction == Reduction::None) {
+    output.resize_({nframe});
+  } else {
+    output = at::tensor(0, input.options());
```
This is wrong.

ezyang commented on aten/src/ATen/native/LossMultiMargin.cpp:
```diff
@@ -0,0 +1,328 @@
+  int64_t nframe, dim;
+  if (ndims <= 1) {
+    nframe = 1;
+    dim = ndims == 0 ? 1 : input.size(0);
```
In a non-slavish port, we should not do this, and not special case the zero-dimensional case. But this is always ok to fix later.

andreaskoepf replied:
ok, you mean it would be better to drop support for scalar-tensors? I saw that the TH impl (accidentially) "supported" it and I did not want to introduce a breaking change.

ezyang replied:
I think the docs are fairly clear that scalars aren't supposed to work. But yeah, I'd leave the breaking change for another PR.

ezyang commented on aten/src/ATen/native/LossMultiMargin.cpp:
```diff
@@ -0,0 +1,328 @@
+
+  if (target.dim() > 0) {
+    for (int64_t i = 0; i < nframe; i++) {
+      int64_t idx = target[i].item().toLong();
```
This is slow!

andreaskoepf committed:
Fix index check performance & segfault of backward impl

@ezyang pointed out that accessing elements with .item() is very
slow. The output of the forward pass is now written using TensorAccessor.
A pointer arithmetic problem was fixed in the backward imlementation that
lead to a segfault.

andreaskoepf commented:
@ezyang Thanks a lot for taking the time of reviewing and finding the performance problem!

The benchmark results now look much better:

### WITH patch / feature branch ###
```
PyTorch version: 1.4.0a0+0249a93
CPU forward 1000 took 3.2876007026061416e-05
CPU forward 10000 took 0.00019442799384705722
CPU forward 100000 took 0.0020630429935408756
CPU forward 1000000 took 0.0190298239904223
CPU forward 10000000 took 0.19668684000498615
CPU forward TOTAL time 0.9072190410079202
CPU for- & backward 1000 took 0.00025180898956023157
CPU for- & backward 10000 took 0.0007014480070210993
CPU for- & backward 100000 took 0.005792188996565528
CPU for- & backward 1000000 took 0.05727226099406835
CPU for- & backward 10000000 took 0.5772217519988772
CPU for- & backward TOTAL time 1.3247482030128594
```

### WITHOUT patch / master ###
```
PyTorch version: 1.4.0a0+4f6b567
CPU forward 1000 took 7.350600208155811e-05
CPU forward 10000 took 0.0006765849975636229
CPU forward 100000 took 0.00676868500886485
CPU forward 1000000 took 0.06623484700685367
CPU forward 10000000 took 0.6726305619959021
CPU forward TOTAL time 1.4349038829968777
CPU for- & backward 1000 took 0.0003252689930377528
CPU for- & backward 10000 took 0.0020270539971534163
CPU for- & backward 100000 took 0.017401589997461997
CPU for- & backward 1000000 took 0.1642533059930429
CPU for- & backward 10000000 took 1.7190422190033132
CPU for- & backward TOTAL time 2.5921302720089443
```

andreaskoepf committed:
Produce scalar loss output for 1D input tensor

The unit test 'MultiMarginLoss_1d_no_reduce' expects a scalar
output for 1D inputs (independent of the dimensionality of the
target vector).

andreaskoepf committed:
Remove old TH (CPU) MultiMarginCriterion

ifedan commented on aten/src/ATen/native/LossMultiMargin.cpp:
```diff
@@ -0,0 +1,337 @@
+      input_data += dim;
+    }
+  } else {
+    scalar_t sum = 0;
```
I think we should use accscalar_t instead

andreaskoepf replied:
What is this type? Is it also useful for CPU code?

andreaskoepf committed:
Use accscalar_t

ezyang commented on aten/src/ATen/native/LossMultiMargin.cpp:
```diff
@@ -0,0 +1,337 @@
+      grad_output,
+      input,
+      target,
+      p.toInt(),
```
This is not anything you have to do. But I do note it's a bit weird that we take a Scalar as argument p, but then immediately turn around and say it must be an int.

andreaskoepf replied:
Actually this results from the fact how this function is defined in native_functions.yaml, where it is: `- func: multi_margin_loss(Tensor self, Tensor target, Scalar p=1, Scalar margin=1, Tensor? weight=None, int reduction=Mean) -> Tensor`. I did not touch this signature - that means the input will be a Scalar. Bu if you look at the documentation and in the python source there is the strange restriction for p to be either 1 or 2 and no other values are allowed. It would be relatively easy to support all power-values - maybe with special fast-cases for 1 and 2. I showed this to @VitalyFedyunin also. For this port I kept the original behavior -> the only thing I added was a check for p==1||p==2 which before was only happening in python, see: https://github.com/pytorch/pytorch/blob/951dd03037d01a4f9ad9703e4f0908e11c59f8c9/torch/nn/functional.py#L2323-L2324

andreaskoepf commented:
@ezyang I think there was a bug in the old argument checks if `input` argument was 1d, e.g. then target could be a LongTensor of arbitrary length which resulted in reading memory outside the input tensor. The following should not be possible:

```
>>> torch.nn.functional.multi_margin_loss(torch.rand(10), torch.tensor([1,2,3]), reduction='none')
tensor(0.9002)
````

After applying this PR the condition is detected and the following error is generated:
```
>>> torch.nn.functional.multi_margin_loss(torch.rand(10), torch.tensor([1,2,3]), reduction='none')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/koepf/code/branches/multi_margin_loss_th_to_aten/torch/nn/functional.py", line 2329, in multi_margin_loss
    return torch._C._nn.multi_margin_loss(input, target, p, margin, weight, reduction_enum)
RuntimeError: inconsistent target size, got: [3]
```
